File: packages/types/src/market-data.ts
/** OHLCV bar for a given timeframe — ts is an ISO 8601 string as returned by the API */
export type OHLCVBar = {
  ts: string; // ISO 8601 UTC string — convert to Unix ms via new Date(ts).getTime()
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
  adjClose: number | null;
};

export type Timeframe = '1m' | '5m' | '15m' | '30m' | '1H' | '4H' | '1D' | '1W' | '1M';

export type Quote = {
  symbol: string;
  price: number | null;
  /** 24-hour price change as a decimal fraction. e.g. 0.03 = +3%. */
  change_24h: number | null;
  volume_24h: number | null;
  /** ISO 8601 UTC string from JSON serialization of datetime. */
  ts: string | null;
};

export type OHLCVResponse = {
  symbol: string;
  timeframe: Timeframe;
  bars: OHLCVBar[];
  source: string;
};

export type BulkQuotesResponse = {
  quotes: Record<string, Quote>;
};

/** Instrument as returned by the REST API */
export type InstrumentResponse = {
  symbol: string;
  name: string;
  asset_class: string;
  exchange: string | null;
  currency: string;
};

export type InstrumentListResponse = {
  instruments: InstrumentResponse[];
  total: number;
  limit: number;
  offset: number;
};

File: services/api/src/schemas/market_data.py
"""
Pydantic response schemas for market data endpoints.

These are the data contracts between the API and its consumers.
Fields match what is stored in ClickHouse (OHLCVRow) and Redis (quote snapshots).
"""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class OHLCVBar(BaseModel):
    """A single OHLCV candlestick bar."""

    ts: datetime = Field(description="Bar open timestamp (UTC).")
    open: float = Field(description="Opening price.")
    high: float = Field(description="Highest price during the interval.")
    low: float = Field(description="Lowest price during the interval.")
    close: float = Field(description="Closing price.")
    volume: float = Field(description="Volume traded during the interval.")
    adj_close: float | None = Field(
        default=None,
        description="Adjusted close. Null for crypto (no corporate actions).",
    )


class OHLCVResponse(BaseModel):
    """Response for GET /market-data/{symbol}/ohlcv."""

    symbol: str = Field(description="Ticker symbol.")
    timeframe: str = Field(description="Candle interval, e.g. '1D', '4H'.")
    bars: list[OHLCVBar] = Field(
        description="Ordered list of OHLCV bars (oldest first)."
    )
    source: str = Field(description="Data provider that sourced these bars.")


class QuoteResponse(BaseModel):
    """Response for GET /market-data/{symbol}/quote."""

    symbol: str = Field(description="Ticker symbol.")
    price: float | None = Field(description="Latest close price. Null if no data yet.")
    change_24h: float | None = Field(
        default=None,
        description="24-hour price change as a decimal fraction. e.g. 0.03 = +3%.",
    )
    volume_24h: float | None = Field(
        default=None, description="24-hour cumulative volume."
    )
    ts: datetime | None = Field(
        default=None, description="Timestamp of the latest bar (UTC)."
    )


class BulkQuotesResponse(BaseModel):
    """Response for GET /market-data/bulk-quotes."""

    quotes: dict[str, QuoteResponse] = Field(
        description="Map of symbol → latest quote. Missing symbols are omitted."
    )
